### PR TITLE
add smoke tests + fix service worker tests

### PR DESCRIPTION
**Pull request by samccone**
https://travis-ci.org/GoogleChrome/lighthouse/jobs/122697303#L1114
^ Interesting... looks like we are not picking up the service worker at all ... curious if anyone has insight here... could it be due to the revert of the force update I did? Is something else going on? 

---

Looks like the force update of service worker API changed a bit today
https://travis-ci.org/GoogleChrome/lighthouse/jobs/122689130#L2606

Which is causing the test to fail until I reverted the commit that enabled this behavior

### CODE REVIEW

**Comment by brendankenny** [comment]
when you wake up, this is now failing :)

**Comment by brendankenny** [comment]
This is a great start, and I agree this starts to test a ton of integration stuff (not to mention driver/gatherers) we don't test at all today.

File: run-smoke-tests.sh [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env bash
+
+cd test/fixtures && python -m SimpleHTTPServer 9999 &
+
+./cli.js http://localhost:9999/online-only.html > results
+
+if ! grep -q "URL responds with a 200 when offline: false" results; then
```
Comment (brendankenny) [inline run-smoke-tests.sh]: would much rather test non-pretty-printed results than strings, as the first is more likely to be stable over time than the second, and easier for humans to quickly scan and keep the test expectations up to date. We can do that in a follow up PR if you want, though.

File: gather-scheduler.js [hunk 2]
```diff
@@ -51,9 +51,6 @@ class GatherScheduler {
       // Clean all browser caches.
       .then(_ => driver.cleanAndDisableBrowserCaches())
 
-      // Force SWs to update on load.
```
Comment (brendankenny) [inline gather-scheduler.js]: @paulirish or @paullewis for what's going on here with the new errors

File: run-smoke-tests.sh (hunk 1 above)
Comment (samccone) [inline run-smoke-tests.sh]: :+1: started doing a version using jq . to parse the response but since it is just an array with objects and things are not keyed inside of an object, the matching logic got much worse and uglier than this.

:+1: to iterating more

File: gather-scheduler.js (hunk 2 above)
Comment (paullewis) [inline gather-scheduler.js]: Checking this out now.

**Comment by paullewis** [comment]
In `run-smoke-tests.sh`:
```diff
+ sleep 5s
```

**Comment by paullewis** [comment]
And switch to ToT with:

```diff
- 'https://download-chromium.appspot.com/dl/Linux_x64?type=continuous'
+ 'https://download-chromium.appspot.com/dl/Linux_x64?type=snapshots'
```

**Comment by samccone** [comment]
so far I have moved to launching chrome directly into the file and removed it from a background process, so nothing is still working. I then went and moved the order of the tests around thinking that it could be due to rerunning the tests in the same chrome instance.. still no good :goat: 

My next steps will be to debug this on my linux desktop and see if i can reproduce the issue. My current thinking is that there may be something weird going on with focus and chrome that is causing the service worker events not to fire... like perhaps since it is on a different (virtual) screen :monkey: it is not even thinking it needs to do any work... but these are just some ramblings... I will need to see if I can repo...

Any other insights that anyone has would be very much appreciated :clock11: :skull:

**Comment by samccone** [comment]
image


via a clean profile ^

note we get started but no actived

**Comment by samccone** [comment]
Just tested with the extension manually, seems like everything is still in working order.

👀 welcome on this one.

**Just a note**, ⚠️ squashing this PR would not be advisable ⚠️

File: test/download-chrome.sh [hunk 3]
```diff
@@ -12,5 +12,5 @@ if [ -e "$LIGHTHOUSE_CHROMIUM_PATH" ]
 then
   echo "cached chrome found"
 else
-  wget 'https://download-chromium.appspot.com/dl/Linux_x64?type=continuous' -O chrome.zip && unzip chrome.zip
+  wget 'https://download-chromium.appspot.com/dl/Linux_x64?type=snapshots' -O chrome.zip && unzip chrome.zip
```
Comment (brendankenny) [inline test/download-chrome.sh]: wasn't the problem not continuous vs snapshots in the end? Can we keep using the continuous build?

File: helpers/browser/driver.js [hunk 4]
```diff
@@ -150,35 +150,33 @@ class ChromeProtocol {
...
+      if (shouldReload) {
+        return this.sendCommand('Page.reload', {ignoreCache: true});
+      }
 
-        return this.sendCommand('Page.navigate', {url});
-      })
-      .then(response => {
+      return this.sendCommand('Page.navigate', {url});
+    }).then(response => {
+      return new Promise((res, _) => {
```
Comment (brendankenny) [inline helpers/browser/driver.js]: `(resolve, reject)`

File: helpers/browser/driver.js [hunk 5]
```diff
@@ -150,35 +150,33 @@ class ChromeProtocol {
...
-        // with a trailing slash on it.
-        //
-        // If the URL matches then we need to issue a reload not navigate
-        // @see https://github.com/GoogleChrome/lighthouse/issues/183
-        const shouldReload = (currentURL === url || currentURL === url + '/');
-        if (shouldReload) {
-          return this.sendCommand('Page.reload', {ignoreCache: true});
-        }
+    return this.sendCommand('Page.enable').then(_ => {
+      return this.sendCommand('Page.getNavigationHistory');
```
Comment (brendankenny) [inline helpers/browser/driver.js]: prefer the original
```
.then(_ => this.sendCommand('Page.getNavigationHistory'))
```
formatting for this line

File: helpers/browser/driver.js [hunk 6]
```diff
@@ -150,35 +150,33 @@ class ChromeProtocol {
...
 
-        return this.sendCommand('Page.navigate', {url});
-      })
-      .then(response => {
+      return this.sendCommand('Page.navigate', {url});
+    }).then(response => {
+      return new Promise((res, _) => {
         this.url = url;
 
         if (!waitForLoaded) {
```
Comment (brendankenny) [inline helpers/browser/driver.js]: why do we even have `waitForLoaded`?

File: gatherers/offline.js [hunk 7]
```diff
@@ -102,22 +102,21 @@ class Offline extends Gather {
...
 
-      return Offline.getOfflinePageStatus(driver)
-        .then(ret => {
-          responseCode = ret.result.value;
-        })
-        .then(_ => Offline.goOnline(driver))
-        .then(_ => {
-          this.artifact = {responseCode};
-        });
+      return Offline.getOfflinePageStatus(driver).then(ret => {
```
Comment (brendankenny) [inline gatherers/offline.js]: this is fine, but does mess up git blame in exchange for arguable readability

File: gather-scheduler.js [hunk 8]
```diff
@@ -18,99 +18,114 @@
...
-  static _runPhase(gatherers, gatherFun) {
-    return gatherers.reduce(
-      (chain, gatherer) => chain.then(_ => gatherFun(gatherer)),
-      Promise.resolve()
-    );
+  static runSeries(fns) {
+    const args = Array.from(arguments).slice(1);
+
+    return fns.reduce((prev, curr) => {
+      return prev.then(_ => curr.apply(null, args));
```
Comment (brendankenny) [inline gather-scheduler.js]: `curr(...args)`

File: test/download-chrome.sh (hunk 3 above)
Comment (samccone) [inline test/download-chrome.sh]: we still need the latest for the new features we are using in 52

File: gather-scheduler.js [hunk 9]
```diff
@@ -18,99 +18,114 @@
...
+      return chain.then(_ => gatherFun(gatherer));
+    }, Promise.resolve());
+  }
 
-      // Disconnect the driver.
-      .then(_ => driver.disconnect())
+  static run(gatherers, options) {
+    const driver = options.driver;
+    const tracingData = {};
+    const self = GatherScheduler;
```
Comment (brendankenny) [inline gather-scheduler.js]: the combination of using `self` to call `static` methods is definitely a code smell. While you're in here, just drop the class and have the static methods be functions

File: gather-scheduler.js [hunk 10]
```diff
@@ -18,99 +18,114 @@
...
 
-      // Gather: tearDown phase.
-      .then(_ => this._runPhase(gatherers,
-        gatherer => gatherer.tearDown(options, tracingData)))
+    if (options.url === undefined || options.url === null) {
```
Comment (brendankenny) [inline gather-scheduler.js]: why would the url be `null`?

File: gather-scheduler.js (hunk 9 above)
Comment (samccone) [inline gather-scheduler.js]: it is just a shorthand for the class, i will rename to klass

Comment (brendankenny) [inline gather-scheduler.js]: that has the same issue

File: gather-scheduler.js [hunk 11]
```diff
@@ -18,99 +18,114 @@
...
+    // Such a hack... since a Page.reload command does not let
+    // a service worker take over we have to trick the browser into going away
+    // and then coming back.
+    return driver.sendCommand('Page.navigate', {url: 'about:blank'}).then(_ => {
+      return driver.gotoURL(options.url, driver.WAIT_FOR_LOADED);
+    });
+  }
 
+  static setupDriver(driver, gatherers, options) {
+    return new Promise((res, _) => {
```
Comment (brendankenny) [inline gather-scheduler.js]: `(resolve, reject)`

Comment (brendankenny) [inline gather-scheduler.js]: or just the original version, returning `driver.beginEmulation()` or `Promise.resolve()` directly

File: gather-scheduler.js [hunk 12]
```diff
@@ -18,99 +18,114 @@
...
-      .then(_ => this._runPhase(gatherers,
-        gatherer => gatherer.tearDown(options, tracingData)))
+    if (options.url === undefined || options.url === null) {
+      throw new Error('You must provide a url to scheduler');
+    }
 
+    return GatherScheduler.runSeries([
+      driver.connect.bind(driver),
+      self.setupDriver,
+      self._phase('setup'),
```
Comment (brendankenny) [inline gather-scheduler.js]: I'm still not wild about this :) Calling the phase methods with a string accessor seems like a symptom of impedance mismatch in two areas:
- `GatherScheduler` is really a collection of static functions but is calling into real gatherer classes, which means you can't just call `runSeries` over the different phases; you need something for the `this` binding
- Pretending these functions are easily composable because they're all endomorphisms over the same type when they actually aren't. Taking the union of domains is one way to force them in that round hole, but it can easily end with an unstructured mess of arguments being passed in (this would be even more obvious if `options` and `tracingData` were split into their individual properties before being passed into `runSeries`). And because our functions aren't side-effect free, you lose information at the call site of what a function is changing, relying on having to inspect the function itself.

In both cases you trade some clarity and static verifiability for an easier to read list of steps. For purity's sake we'd want to make them actually endomorphisms/monadic, instead of having the state repeatedly reinjected, or if it's the readability we're after we actually want `async`/`await` here. The first approach is probably overkill, the second isn't possible yet.

I personally feel that something closer to the existing `then`-heavy structure, with the driver steps pulled out as you have here, is a better expression of what we're doing here while we're waiting on `await` support to drop, rather than making a pipeline over (conceptually-)orthogonal domains.

Comment (brendankenny) [inline gather-scheduler.js]: As an example of why the second bullet point isn't ideal: I've been thinking about nice ways we can get away from the magic `this.artifact` on gatherers, returning a promise that resolves to the artifact at some stage. This would require adding an `artifacts` array to the arguments passed into `runSeries`, which means future readers would have to go through all the called functions to find out that only calls to `runPhase` was creating possible artifacts.

Comment (brendankenny) [inline gather-scheduler.js]: This is, incidentally, why I was hoping for separate PRs for the addition of reload phases vs the scheduler refactor :)

File: helpers/browser/driver.js (hunk 6 above)
Comment (samccone) [inline helpers/browser/driver.js]: duno, let's open an issue and take a look

File: gather-scheduler.js (hunk 12 above)
Comment (samccone) [inline gather-scheduler.js]: we can iterate on this more, I felt that this was an OK first step.. you are correct none of these are side effect free or re-entrant (however I do not think these can ever be since we do not have state free exits). The goal here was the make the list of steps readable, and I feel pretty ok with what this ended up being like.

I felt that adding the new steps onto the .then chain pushed it over the limit of readability.


I want to iterate more on this and we can continue to tweak

Comment (samccone) [inline gather-scheduler.js]: you lose information at the call site of what a function is changing, relying on having to inspect the function itself

^ I think it was a lot like this before, it was all just a really long .then chain and hard to grok what was going on and what the flow was.

Comment (samccone) [inline gather-scheduler.js]: > which means future readers would have to go through all the called functions to find out that only calls to runPhase was creating possible artifacts.

yep, it is an abstraction, I personally am willing to pay this cost for the current layout of the flow code.

**Comment by samccone** [comment]
Let me know what you want me to do, I can revert all of this if you think this is a no go.

File: gather-scheduler.js (hunk 12 above)
Comment (samccone) [inline gather-scheduler.js]: > This is, incidentally, why I was hoping for separate PRs for the addition of reload phases vs the scheduler refactor :)

I did individual commits to make moving / reverting easy

as I said above I found layering on more .then clauses unacceptable

File: gatherers/offline.js (hunk 7 above)
Comment (samccone) [inline gatherers/offline.js]: ack

File: helpers/browser/driver.js (hunk 4 above)
Comment (samccone) [inline helpers/browser/driver.js]: done

File: helpers/browser/driver.js (hunk 5 above)
Comment (samccone) [inline helpers/browser/driver.js]: done

File: gather-scheduler.js (hunk 8 above)
Comment (samccone) [inline gather-scheduler.js]: nice 👍

File: gather-scheduler.js (hunk 9 above)
Comment (samccone) [inline gather-scheduler.js]: ok let's do this in another PR seems like a good abstraction / improvement.. or at least one that we can play around with in another PR sandbox

File: gather-scheduler.js (hunk 10 above)
Comment (samccone) [inline gather-scheduler.js]: unknown, just handling the case :)

I did not wanted to be explicit here because if someone passed `''` I would not want to fail

File: gather-scheduler.js (hunk 11 above)
Comment (samccone) [inline gather-scheduler.js]: since there is some logic inside of the block here I wanted to wrap it all in a promise constructor to handle the case where something could potentially throw 💥

**Comment by paullewis** [comment]
The main issues here are around the refactor of the gatherer. Can we land this with the current gatherer logic (with the addition of the second page load) and deal with a gatherer refactor in a separate PR? I would like the SW tests to be fixed sooner rather than later.

File: gather-scheduler.js [hunk 13]
```diff
@@ -18,99 +18,124 @@
...
 
-      // Gather: tearDown phase.
-      .then(_ => this._runPhase(gatherers,
-        gatherer => gatherer.tearDown(options, tracingData)))
+    if (options.url === undefined || options.url === null) {
+      throw new Error('You must provide a url to scheduler');
+    }
 
+    return driver.connect().then(_ => {
+      return GatherScheduler.setupDriver(driver, gatherers, options);
```
Comment (paullewis) [inline gather-scheduler.js]: Why the explicit returns?

Comment (samccone) [inline gather-scheduler.js]: multi-line =>

Comment (paullewis) [inline gather-scheduler.js]: Yeah, but whyyyyyyy?

Comment (samccone) [inline gather-scheduler.js]: lines were too long

Comment (paullewis) [inline gather-scheduler.js]: 👍

Comment (brendankenny) [inline gather-scheduler.js]: that's why I was saying make the methods just functions (since they're already operating as ones). Then you can just do, like,
```
    return driver.connect()
      .then(_ => setupDriver(driver, gatherers, options))
      .then(_ => runPhase(gatherers, gatherer => gatherer.setup(options)))
      .then(_ => beginPassiveCollection(driver))
      .then(_ => runPhase(gatherers, gatherer => gatherer.beforePageLoad(options)))
      .then(_ => loadPage(driver, gatherers, options))
      .then(_ => runPhase(gatherers, gatherer => gatherer.afterPageLoad(options)))
      .then(_ => endPassiveCollection(driver, tracingData))
      .then(_ => runPhase(gatherers,
          gatherer => gatherer.afterTraceCollected(options, tracingData)))
      .then(_ => runPhase(gatherers, gatherer => gatherer.reloadSetup(options)))
      .then(_ => runPhase(gatherers, gatherer => gatherer.beforeReloadPageLoad(options)))
      .then(_ => reloadPage(driver, options))
      .then(_ => runPhase(gatherers, gatherer => gatherer.afterReloadPageLoad(options)))
      .then(_ => driver.disconnect())
      .then(_ => runPhase(gatherers, gatherer => gatherer.tearDown(options)))
      .then(_ => {
        // Collate all the gatherer results.
        return gatherers.map(g => g.artifact).concat(
            {networkRecords: tracingData.networkRecords},
            {traceContents: tracingData.traceContents});
      });
```

File: gather-scheduler.js [hunk 14]
```diff
@@ -18,99 +18,124 @@
 
 class GatherScheduler {
 
-  static _runPhase(gatherers, gatherFun) {
-    return gatherers.reduce(
-      (chain, gatherer) => chain.then(_ => gatherFun(gatherer)),
-      Promise.resolve()
-    );
+  static runSeries(fns) {
+    const args = Array.from(arguments).slice(1);
```
Comment (brendankenny) [inline gather-scheduler.js]: it'll be so nice when we get rest parameters in node as well

Comment (brendankenny) [inline gather-scheduler.js]: I think you want to remove this, actually, since you aren't using it in the current changes

File: audits/mobile-friendly/display.js [hunk 15]
```diff
@@ -55,10 +55,13 @@ class ManifestDisplay extends Audit {
...
     const manifest = artifacts.manifest.value;
+    const displayValue = (!manifest || !manifest.display) ? undefined : manifest.display.value;
+
+    const hasRecommendedValue = ManifestDisplay.hasRecommendedValue(displayValue);
 
     return ManifestDisplay.generateAuditResult(
-      ManifestDisplay.hasRecommendedValue(manifest.display.value),
-      manifest.display.value,
+      hasRecommendedValue,
+      displayValue,
```
Comment (brendankenny) [inline audits/mobile-friendly/display.js]: the manifest parser also has a debug string for display value if it isn't one of `fullscreen`, `standalone`, `minimal-ui`, or `browser` it might be worth piping through. Not sure the best way to combine it with this warning, though.

File: gather-scheduler.js (hunk 13 above)
Comment (brendankenny) [inline gather-scheduler.js]: just to be clear, this can be held off for a future PR, though :)

File: test/download-chrome.sh (hunk 3 above)
Comment (brendankenny) [inline test/download-chrome.sh]: 👍

File: audits/mobile-friendly/display.js (hunk 15 above)
Comment (samccone) [inline audits/mobile-friendly/display.js]: open up an issue to track we can improve it there

File: gather-scheduler.js (hunk 13 above)
Comment (samccone) [inline gather-scheduler.js]: we can iterate on it more

File: gather-scheduler.js (hunk 14 above)
Comment (samccone) [inline gather-scheduler.js]: good catch, fixed

File: gather-scheduler.js (hunk 13 above)
Comment (paulirish) [inline gather-scheduler.js]: Line length. :3

**Comment by brendankenny** [comment]
Nice! 💣 => 💥 

I'm sorry I make all things painful 🐗
